File: src/queries/pageviews.ts
import { rawQuery, buildFilterSQL, type Filter } from "@/lib/queries";

export interface PageMetrics {
  urlPath: string;
  pageTitle: string | null;
  views: number;
  visitors: number;
  avgTimeOnPage: number | null;
  medianTimeOnPage: number | null;
  entries: number;
  exits: number;
  bounceRate: number;
}

/**
 * Get detailed page metrics including time on page and entry/exit data.
 * When pageMode is "slug", strips query strings and normalizes trailing slashes.
 */
export async function getPageMetrics(
  websiteId: string,
  startDate: Date,
  endDate: Date,
  limit = 50,
  offset = 0,
  filters: Filter[] = [],
  pageMode: "path" | "slug" = "path"
): Promise<PageMetrics[]> {
  const { sql: filterWhere, params: filterParams, needsSessionJoin } =
    buildFilterSQL(filters);

  // In the CTE we use "we" alias for consistency with filter SQL
  const sessionJoinCTE = needsSessionJoin
    ? "JOIN session s ON s.session_id = we.session_id"
    : "";

  // In slug mode, normalize: strip trailing slashes, collapse to lowercase path only
  const urlExpr =
    pageMode === "slug"
      ? "REGEXP_REPLACE(SPLIT_PART(we.url_path, '?', 1), '/+$', '')"
      : "we.url_path";
  // Ensure empty slug (root with trailing slash stripped) becomes '/'
  const slugCoalesce =
    pageMode === "slug"
      ? `CASE WHEN ${urlExpr} = '' THEN '/' ELSE ${urlExpr} END`
      : urlExpr;

  const results = await rawQuery<{
    url_path: string;
    page_title: string | null;
    views: bigint;
    visitors: bigint;
    avg_time_on_page: number | null;
    median_time_on_page: number | null;
    entries: bigint;
    exits: bigint;
    bounce_rate: number;
  }>(
    `WITH filtered_events AS (
      SELECT ${slugCoalesce} AS url_path, we.page_title, we.visit_id, we.session_id, we.created_at
      FROM website_event we
      ${sessionJoinCTE}
      WHERE we.website_id = {{websiteId::uuid}}
        AND we.created_at BETWEEN {{startDate::timestamptz}} AND {{endDate::timestamptz}}
        AND we.event_type = 1
        ${filterWhere}
    ),
    page_sequence AS (
      SELECT
        url_path,
        page_title,
        visit_id,
        session_id,
        created_at,
        LEAD(created_at) OVER (PARTITION BY visit_id ORDER BY created_at) AS next_page_at,
        ROW_NUMBER() OVER (PARTITION BY visit_id ORDER BY created_at ASC) AS rn_entry,
        ROW_NUMBER() OVER (PARTITION BY visit_id ORDER BY created_at DESC) AS rn_exit
      FROM filtered_events
    ),
    visit_bounces AS (
      SELECT visit_id, COUNT(*) AS page_count
      FROM filtered_events
      GROUP BY visit_id
    )
    SELECT
      ps.url_path,
      MAX(ps.page_title) AS page_title,
      COUNT(*)::bigint AS views,
      COUNT(DISTINCT ps.session_id)::bigint AS visitors,
      AVG(EXTRACT(EPOCH FROM (ps.next_page_at - ps.created_at)))
        FILTER (WHERE ps.next_page_at IS NOT NULL) AS avg_time_on_page,
      PERCENTILE_CONT(0.5) WITHIN GROUP (
        ORDER BY EXTRACT(EPOCH FROM (ps.next_page_at - ps.created_at))
      ) FILTER (WHERE ps.next_page_at IS NOT NULL) AS median_time_on_page,
      COUNT(*) FILTER (WHERE ps.rn_entry = 1)::bigint AS entries,
      COUNT(*) FILTER (WHERE ps.rn_exit = 1)::bigint AS exits,
      CASE
        WHEN COUNT(*) FILTER (WHERE ps.rn_entry = 1) = 0 THEN 0
        ELSE (
          COUNT(*) FILTER (WHERE ps.rn_entry = 1 AND vb.page_count = 1)::float /
          NULLIF(COUNT(*) FILTER (WHERE ps.rn_entry = 1), 0) * 100
        )
      END AS bounce_rate
    FROM page_sequence ps
    LEFT JOIN visit_bounces vb ON ps.visit_id = vb.visit_id
    GROUP BY ps.url_path
    ORDER BY views DESC
    LIMIT ${limit} OFFSET ${offset}`,
    { websiteId, startDate, endDate, ...filterParams }
  );

  return results.map((r) => ({
    urlPath: r.url_path,
    pageTitle: r.page_title,
    views: Number(r.views),
    visitors: Number(r.visitors),
    avgTimeOnPage: r.avg_time_on_page,
    medianTimeOnPage: r.median_time_on_page,
    entries: Number(r.entries),
    exits: Number(r.exits),
    bounceRate: r.bounce_rate ?? 0,
  }));
}

export interface PageTimeSeries {
  time: string;
  views: number;
  visitors: number;
}

/**
 * Get time series for a specific page.
 */
export async function getPageTimeSeries(
  websiteId: string,
  urlPath: string,
  startDate: Date,
  endDate: Date,
  granularity: string,
  filters: Filter[] = []
): Promise<PageTimeSeries[]> {
  const validGranularities = ["minute", "hour", "day", "week", "month"];
  const gran = validGranularities.includes(granularity) ? granularity : "day";
  const { sql: filterWhere, params: filterParams, needsSessionJoin } =
    buildFilterSQL(filters);

  const sessionJoin = needsSessionJoin
    ? "JOIN session s ON s.session_id = we.session_id"
    : "";

  const results = await rawQuery<{
    time: Date;
    views: bigint;
    visitors: bigint;
  }>(
    `SELECT
      date_trunc('${gran}', we.created_at) AS time,
      COUNT(*)::bigint AS views,
      COUNT(DISTINCT we.session_id)::bigint AS visitors
    FROM website_event we
    ${sessionJoin}
    WHERE we.website_id = {{websiteId::uuid}}
      AND we.created_at BETWEEN {{startDate::timestamptz}} AND {{endDate::timestamptz}}
      AND we.event_type = 1
      AND we.url_path = {{urlPath}}
      ${filterWhere}
    GROUP BY 1
    ORDER BY 1 ASC`,
    { websiteId, startDate, endDate, urlPath, ...filterParams }
  );

  return results.map((r) => ({
    time: r.time instanceof Date ? r.time.toISOString() : String(r.time),
    views: Number(r.views),
    visitors: Number(r.visitors),
  }));
}

File: core/mantecato_core/queries/pageviews.py
"""Page-level analytics queries — aggregate pageview metrics per URL.

Privacy-first: only pageview counts per URL. No visitors, sessions, bounce rates,
time-on-page, entry/exit, or navigation tracking (those require persistent identifiers).
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from core.mantecato_core.database import raw_query
from core.mantecato_core.filters import Filter, GRANULARITIES, prepare_filters, safe_identifier
from core.mantecato_core.queries.orm_fallbacks import (
    pageview_queryset,
    pageview_time_series_rows,
    should_use_orm_fallback,
)


def get_page_metrics(
    website_id: str,
    start_date: datetime,
    end_date: datetime,
    limit: int = 50,
    offset: int = 0,
    filters: list[Filter] | None = None,
    page_mode: str = "path",
) -> list[dict[str, Any]]:
    """Compute aggregate per-page pageview counts with pagination.

    Returns each tracked URL with its view count and most recent page title.
    Visitor/session/bounce/time metrics are not available because they require
    persistent identifiers.

    Args:
        website_id: UUID of the tracked website.
        start_date: Inclusive start of the analysis window.
        end_date: Exclusive end of the analysis window.
        limit: Maximum number of page rows per page of results (default 50).
        offset: Row offset for pagination (default 0).
        filters: Optional list of column filters to narrow the dataset.
        page_mode: URL normalization mode -- ``"path"`` or ``"slug"``.

    Returns:
        List of dicts with ``urlPath``, ``pageTitle``, ``views``, sorted by
        views descending.
    """
    if should_use_orm_fallback():
        from django.db.models import Count, Max

        rows = (
            pageview_queryset(website_id, start_date, end_date, filters)
            .values("url_path")
            .annotate(page_title=Max("page_title"), views=Count("event_id"))
            .order_by("-views", "url_path")[offset : offset + limit]
        )
        return [
            {
                "urlPath": row["url_path"] or "/",
                "pageTitle": row["page_title"],
                "views": int(row["views"] or 0),
            }
            for row in rows
        ]

    filters = filters or []
    filter_where, filter_params, _ = prepare_filters(filters)

    if page_mode == "slug":
        url_expr = "REGEXP_REPLACE(SPLIT_PART(we.url_path, '?', 1), '/+$', '')"
        slug_coalesce = f"CASE WHEN {url_expr} = '' THEN '/' ELSE {url_expr} END"
    else:
        slug_coalesce = "we.url_path"

    rows = raw_query(
        f"""SELECT
      {slug_coalesce} AS url_path,
      MAX(we.page_title) AS page_title,
      COUNT(*)::bigint AS views
    FROM website_event we
    WHERE we.website_id = {{websiteId::uuid}}
      AND we.created_at BETWEEN {{startDate::timestamptz}} AND {{endDate::timestamptz}}
      AND we.event_type = 1
      {filter_where}
    GROUP BY url_path
    ORDER BY views DESC
    LIMIT {limit} OFFSET {offset}""",
        {
            "websiteId": website_id,
            "startDate": start_date,
            "endDate": end_date,
            **filter_params,
        },
    )

    return [
        {
            "urlPath": row["url_path"],
            "pageTitle": row["page_title"],
            "views": int(row["views"] or 0),
        }
        for row in rows
    ]


def get_page_time_series(
    website_id: str,
    url_path: str,
    start_date: datetime,
    end_date: datetime,
    granularity: str,
    filters: list[Filter] | None = None,
) -> list[dict[str, Any]]:
    """Generate a time series of pageviews for a specific page."""
    if should_use_orm_fallback():
        scoped_filters = [*(filters or []), Filter("url_path", "eq", url_path)]
        rows = pageview_time_series_rows(
            website_id,
            start_date,
            end_date,
            safe_identifier(granularity, GRANULARITIES, "day"),
            scoped_filters,
        )
        return [{"time": row["time"], "views": row["pageviews"]} for row in rows]

    filters = filters or []
    filter_where, filter_params, _ = prepare_filters(filters)
    gran = granularity if granularity in ("minute", "hour", "day", "week", "month") else "day"

    rows = raw_query(
        f"""SELECT
      date_trunc('{gran}', we.created_at) AS time,
      COUNT(*)::bigint AS views
    FROM website_event we
    WHERE we.website_id = {{websiteId::uuid}}
      AND we.created_at BETWEEN {{startDate::timestamptz}} AND {{endDate::timestamptz}}
      AND we.event_type = 1
      AND we.url_path = {{urlPath}}
      {filter_where}
    GROUP BY 1
    ORDER BY 1 ASC""",
        {
            "websiteId": website_id,
            "startDate": start_date,
            "endDate": end_date,
            "urlPath": url_path,
            **filter_params,
        },
    )

    return [
        {
            "time": row["time"].isoformat()
            if isinstance(row["time"], datetime)
            else str(row["time"]),
            "views": int(row["views"] or 0),
        }
        for row in rows
    ]
